cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java

```diff
@@ -781,4 +790,102 @@ private void writeSystemMetadata(Id.Artifact artifactId, ArtifactInfo artifactIn
     ArtifactSystemMetadataWriter writer = new ArtifactSystemMetadataWriter(metadataStore, artifactId, artifactInfo);
     writer.write();
   }
+
+  /**
+   * Given plugin artifact, find the parent artifact of plugin, after creating classloader with parent artifact
+   * and plugin, we invoke the plugin method in the plugin identified by type and name and return the response.
+   * @param namespace namespace
+   * @param artifactId plugin artifact id
+   * @param pluginType type of the plugin
+   * @param pluginName name of the plugin
+   * @param methodName name of the method
+   * @param parameters request body which corresponds to method parameter
+   * @return Serialized string response from method invocation
+   * @throws ArtifactNotFoundException
+   * @throws IOException
+   * @throws IllegalArgumentException
+   * @throws BadRequestException
+   * @throws ClassNotFoundException
+   * @throws InvocationTargetException
+   */
+  public String pickParentArtifactAndInvokePluginEndpoint(Id.Namespace namespace,
+                                                          Id.Artifact artifactId, String pluginType,
+                                                          String pluginName, String methodName,
+                                                          String parameters) throws ArtifactNotFoundException,
+    IOException, IllegalArgumentException, BadRequestException, ClassNotFoundException, InvocationTargetException {
+    ArtifactDetail artifactDetail = artifactStore.getArtifact(artifactId);
+    return invokePluginEndpoint(namespace, artifactDetail, pluginType, pluginName,
+                                pickParentArtifact(artifactDetail), methodName, parameters);
+  }
+
+  private ArtifactDescriptor pickParentArtifact(ArtifactDetail artifactDetail)
+    throws InvalidArtifactException, ArtifactNotFoundException, IOException {
+
+    // get parent artifacts
+    Set<ArtifactRange> parentArtifactRanges = artifactDetail.getMeta().getUsableBy();
+    if (parentArtifactRanges.isEmpty()) {
+      throw new IllegalArgumentException("No parent artifact found for the plugin artifact");
+    }
+
+    // just pick the first parent artifact from the set.
+    ArtifactRange parentArtifactRange = parentArtifactRanges.iterator().next();
+    List<ArtifactDetail> artifactDetails = artifactStore.getArtifacts(parentArtifactRange);
+    if (artifactDetails.isEmpty()) {
+      // should not happen
+      throw new InvalidArtifactException(String.format("No Artifact found in the range %s", parentArtifactRange));
```

Comment:
throw an ArtifactNotFoundException saying the parent artifact could not be found